### PR TITLE
Add momentum

### PR DESCRIPTION
**Pull request by jeongseok-meta**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by jeongseok-meta** [comment]
@conda-forge/help-c-cpp, ready for review!

File: recipes/Momentum/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+source:
+  - url: https://github.com/facebookincubator/momentum/archive/v{{ version }}.tar.gz
+    sha256: f37bb8288b873de3b804c367eadeec81c02bff2ceff470f35a9427ed4b426bfe
+    patches:
+      - specify-output-dirs.patch
+      - include-cstdlib.patch
+
+build:
+  number: 0
+  skip: true  # [linux or osx]
```
Comment (xhochy) [inline recipes/Momentum/meta.yaml:17]: Is this broad skip intentional?

File: recipes/Momentum/build.sh [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  # See https://conda-forge.org/docs/maintainer/knowledge_base.html#newer-c-features-with-old-sdk
+  CXXFLAGS="${CXXFLAGS} -D_LIBCPP_DISABLE_AVAILABILITY"
+fi
+
+cmake $SRC_DIR \
+  ${CMAKE_ARGS} \
+  -G Ninja \
+  -B build \
+  -DCMAKE_INSTALL_PREFIX=$PREFIX \
+  -DCMAKE_PREFIX_PATH=$PREFIX \
```
Comment (xhochy) [inline recipes/Momentum/build.sh:15]: Part of `${CMAKE_ARGS}`
```suggestion
```

File: recipes/Momentum/build.sh [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  -DCMAKE_BUILD_TYPE=Release \
+  -DCMAKE_INSTALL_LIBDIR=lib \
```
Comment (xhochy) [inline recipes/Momentum/build.sh:17]: Part of `${CMAKE_ARGS}`
```suggestion
```

**Review by xhochy — CHANGES_REQUESTED** [review]
Is this a header-only library?

File: recipes/Momentum/meta.yaml (hunk 1 above)
Comment (jeongseok-meta) [reply recipes/Momentum/meta.yaml:17]: Yes, I would like to enable them after it's converted to a feedstock repo for faster iteration, if that's acceptable :)

**Comment by jeongseok-meta** [comment]
> Is this a header-only library?

It's not: https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=969563&view=logs&j=240f1fee-52bc-5498-a14a-8361bde76ba0&t=7c0f8eae-6d6f-51bf-636f-73a1a7fb1bc4&l=2343

**Comment by jeongseok-meta** [comment]
@xhochy, I have addressed your comment and am ready for a second review. Thank you!

File: recipes/Momentum/meta.yaml (hunk 1 above)
Comment (xhochy) [reply recipes/Momentum/meta.yaml:17]: If you enable them here, I would be able to help you a bit. It's surprising to see (like the first time) that someone got something to build on Windows but fails on the unixes.

**Comment by xhochy** [comment]
> > Is this a header-only library?
> 
> It's not: https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=969563&view=logs&j=240f1fee-52bc-5498-a14a-8361bde76ba0&t=7c0f8eae-6d6f-51bf-636f-73a1a7fb1bc4&l=2343



> > Is this a header-only library?
> 
> It's not: https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=969563&view=logs&j=240f1fee-52bc-5498-a14a-8361bde76ba0&t=7c0f8eae-6d6f-51bf-636f-73a1a7fb1bc4&l=2343

Then please make this output shared libraries (especially on Unix) and add a run_exports.

**Comment by jeongseok-meta** [comment]
@conda-forge-admin, please restart ci

**Comment by jeongseok-meta** [comment]
[Build failure on linux64](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=970671&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=892):

```
CMake Error at /home/conda/staged-recipes/build_artifacts/momentum_1720277066076/_build_env/share/cmake-3.30/Modules/FindPackageHandleStandardArgs.cmake:233 (message):
  Could NOT find Python3 (missing: Python3_INCLUDE_DIRS Python3_LIBRARIES
  Development Development.Module Development.Embed) (found version "3.12.4")
```

It fails to find Python3 by `find_package(Python3 REQUIRED COMPONENTS Interpreter Development)` when `python` is added to build dependency in `meta.yaml`:

https://github.com/conda-forge/staged-recipes/pull/26741/files#diff-ca3ef0eb152892ce9d7a357e85c3068cd445fdeb4b40db658c93caa773e2c7f2R23

**Comment by jeongseok-meta** [comment]
[Build failure on osx64](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=970671&view=logs&j=e35d4f76-8ff2-5536-d795-df91e63eb9f7&t=fa7b4b17-b6ff-5c9c-8cfc-15f888c92310&l=1585):

```
FAILED: CMakeFiles/simd_constraints.dir/momentum/character_solver/simd_normal_error_function.cpp.o 
$BUILD_PREFIX/bin/x86_64-apple-darwin13.4.0-clang++ -DFMT_SHARED -DMOMENTUM_WITH_SPDLOG=1 -DSPDLOG_COMPILED_LIB -DSPDLOG_FMT_EXTERNAL -DSPDLOG_SHARED_LIB -Dsimd_constraints_EXPORTS -I$SRC_DIR -I$SRC_DIR/build -I$SRC_DIR/axel -I$SRC_DIR/build/axel -isystem $PREFIX/include/eigen3 -isystem $PREFIX/include/dispenso/third-party/moodycamel -march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fstack-protector-strong -O2 -pipe -stdlib=libc++ -fvisibility-inlines-hidden -fmessage-length=0 -isystem $PREFIX/include -fdebug-prefix-map=$SRC_DIR=/usr/local/src/conda/momentum-0.1.1 -fdebug-prefix-map=$PREFIX=/usr/local/src/conda-prefix -D_LIBCPP_DISABLE_AVAILABILITY -O3 -DNDEBUG -isysroot /Applications/Xcode_14.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk -mmacosx-version-min=10.13 -fPIC -march=native -MD -MT CMakeFiles/simd_constraints.dir/momentum/character_solver/simd_normal_error_function.cpp.o -MF CMakeFiles/simd_constraints.dir/momentum/character_solver/simd_normal_error_function.cpp.o.d -o CMakeFiles/simd_constraints.dir/momentum/character_solver/simd_normal_error_function.cpp.o -c $SRC_DIR/momentum/character_solver/simd_normal_error_function.cpp
In file included from $SRC_DIR/momentum/character_solver/simd_normal_error_function.cpp:8:
In file included from $SRC_DIR/momentum/character_solver/simd_normal_error_function.h:13:
$SRC_DIR/momentum/common/aligned.h:42:15: error: reference to unresolved using declaration
  return std::aligned_alloc(align, size);
              ^
$BUILD_PREFIX/bin/../include/c++/v1/cstdlib:148:1: note: using declaration annotated with 'using_if_exists' here
using ::aligned_alloc _LIBCPP_USING_IF_EXISTS;
^
In file included from $SRC_DIR/momentum/character_solver/simd_normal_error_function.cpp:8:
In file included from $SRC_DIR/momentum/character_solver/simd_normal_error_function.h:13:
$SRC_DIR/momentum/common/aligned.h:42:10: error: excess elements in scalar initializer
  return std::aligned_alloc(align, size);
         ^                       ~~~~~~
2 errors generated.
```

[Building it in the upstream CI works fine](https://github.com/facebookincubator/momentum/actions/workflows/ci_macos.yml).

File: recipes/Momentum/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  version: {{ version }}
+
+source:
+  - url: https://github.com/facebookincubator/momentum/archive/v{{ version }}.tar.gz
+    sha256: f37bb8288b873de3b804c367eadeec81c02bff2ceff470f35a9427ed4b426bfe
+
+build:
+  number: 0
+  run_exports:
+    - {{ pin_compatible(name, max_pin='x.x') }}
```
Comment (xhochy) [inline recipes/Momentum/meta.yaml:15]: Not sure why the linter suggest `pin_compatible`, I have only ever seen `pin_subpackage` here.
```suggestion
    - {{ pin_subpackage(name, max_pin='x.x') }}
```

File: recipes/Momentum/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+requirements:
+  build:
+    - {{ compiler('cxx') }}
+    - {{ stdlib('c') }}
+    - cmake
+    - ninja  # [not win]
+    - python
```
Comment (xhochy) [inline recipes/Momentum/meta.yaml:23]: This needs to be in host

**Review by xhochy — CHANGES_REQUESTED** [review]
For `osx`, you will need to add a conda_build_config.yaml with

```
c_stdlib_version:  # [osx and x86_64]
 - '10.15'  # [osx and x86_64]
```

This will use a newer macOS deployment target that should have aligned_alloc.

File: recipes/Momentum/meta.yaml (hunk 4 above)
Comment (jeongseok-meta) [reply recipes/Momentum/meta.yaml]: Hmm, I'm not sure. The linter still suggests using `pin_compatible`, and the documentation doesn't explain the differences between them:

https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#conda-build-specific-jinja2-functions

**Comment by jeongseok-meta** [comment]
It seems that it builds multiple times for different python versions (as `python` is added in to `host` and `run`). However, `python` is only needed during the build process. Is there a more efficient way to handle this instead of adding `python` to both the host and run dependencies?

**Comment by jeongseok-meta** [comment]
@ xhochy I have addressed your comments and this pull request is now ready for another review. Thank you for helping to enable building for Linux and macOS!

**Review by xhochy — APPROVED** [review]

**Pull request merged**